File: output/简单易懂的方法解决K个一组翻转链表/简单易懂的方法解决K个一组翻转链表.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
 
struct ListNode* reverseKGroup(struct ListNode* head, int k) {
	 struct ListNode* pre, * p;
	 struct ListNode *rear, * curhead,*curp;
	 int i;

	 curhead = (struct ListNode*)malloc(sizeof(struct ListNode));
	 pre = p = head;
	 while (p)
	 {
		 for (i = 1; i < k; i++)
		 {
			 if (!p)
				 break;
			 p = p->next;
		 }
		 if (p)
		 {
			 rear = p->next;
			 curhead->next = NULL;
			 if (pre == head)
				 head = curhead;
			 while (pre != rear)
			 {
				 curp = pre;
				 pre = pre->next;
				 curp->next = curhead->next;
				 curhead->next = curp;
			 }
			 while (curhead->next)
			 {
				 curhead = curhead->next;
			 }
			 curhead->next = rear;
			 p = pre;
		 }

	 }
	 return head->next;
 }

File: output/简单易懂的方法解决K个一组翻转链表/简单易懂的方法解决K个一组翻转链表.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode {
            next: None,
            val,
        }
    }
}

pub fn reverse_k_group(head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
    if head.is_none() || k <= 1 {
        return head;
    }
    
    let mut dummy = Box::new(ListNode::new(0));
    dummy.next = head;
    let mut prev_group_end = &mut dummy as *mut Box<ListNode>;
    
    loop {
        // Find the k-th node from current position
        let mut kth_node = unsafe { &mut *prev_group_end };
        for _ in 0..k {
            if kth_node.next.is_none() {
                return dummy.next;
            }
            kth_node = kth_node.next.as_mut().unwrap();
        }
        
        let next_group_start = kth_node.next.take();
        
        // Reverse the current group
        let group_start = unsafe { (*prev_group_end).next.take() };
        let (new_group_start, new_group_end) = reverse_list(group_start);
        
        // Connect the reversed group
        unsafe {
            (*prev_group_end).next = new_group_start;
            (*new_group_end).next = next_group_start;
        }
        
        // Move prev_group_end to the end of current group
        prev_group_end = new_group_end;
    }
}

fn reverse_list(head: Option<Box<ListNode>>) -> (Option<Box<ListNode>>, *mut Box<ListNode>) {
    let mut prev = None;
    let mut current = head;
    let mut tail_ref: *mut Box<ListNode> = std::ptr::null_mut();
    
    while let Some(mut node) = current {
        if tail_ref.is_null() {
            tail_ref = &mut node as *mut Box<ListNode>;
        }
        current = node.next.take();
        node.next = prev;
        prev = Some(node);
    }
    
    (prev, tail_ref)
}